File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CreatingHashAlgorithms.vb
Imports System.Security.Cryptography

Namespace Tests.Diagnostics

    Public Class InsecureHashAlgorithm

        Public Sub Hash(temp as Byte())
            Dim DSACng = new DSACng(10)
'                        ^^^^^^^^^^^^^^ {{Make sure this weak hash algorithm is not used in a sensitive context here.}}
            Dim DSACryptoServiceProvider = new DSACryptoServiceProvider() ' Noncompliant
            Dim DSACreate = DSA.Create() ' Noncompliant
            Dim DSACreateWithParam = DSA.Create("DSA") ' Noncompliant
            Dim DSACreateFromName = CryptoConfig.CreateFromName("DSA") ' Noncompliant
            Dim DSAAsymmetricAlgorithm = AsymmetricAlgorithm.Create("DSA") ' Noncompliant

            Dim HMACCreate = HMAC.Create() ' Noncompliant
            Dim HMACCreateWithParam = HMAC.Create("HMACMD5") ' Noncompliant

            Dim HMACMD5 = new HMACMD5() ' Noncompliant
            Dim HMACMD5Create = HMACMD5.Create() ' Noncompliant
            Dim HMACMD5CreateWithParam = HMACMD5.Create("HMACMD5") ' Noncompliant
            Dim HMACMD5KeyedHashAlgorithm = KeyedHashAlgorithm.Create("HMACMD5") ' Noncompliant
            Dim HMACMD5CryptoConfig = CryptoConfig.CreateFromName("HMACMD5") ' Noncompliant

            Dim HMACSHA1 = new HMACSHA1() ' Noncompliant
            Dim HMACSHA1Create = HMACMD5.Create() ' Noncompliant
            Dim HMACSHA1CreateWithParam = HMACMD5.Create("HMACSHA1") ' Noncompliant
            Dim HMACSHA1KeyedHashAlgorithm = KeyedHashAlgorithm.Create("HMACSHA1") ' Noncompliant
            Dim HMACSHA1CryptoConfig = CryptoConfig.CreateFromName("HMACSHA1") ' Noncompliant

            Dim HMACSHA256Create = HMACSHA256.Create("HMACSHA256")
            Dim HMACSHA256KeyedHashAlgorithm = KeyedHashAlgorithm.Create("HMACSHA256")
            Dim HMACSHA256CryptoConfig = CryptoConfig.CreateFromName("HMACSHA256")

            Dim MD5CryptoServiceProvider = new MD5CryptoServiceProvider() ' Noncompliant
            Dim MD5CryptoConfig = CryptoConfig.CreateFromName("MD5") ' Noncompliant
            Dim MD5HashAlgorithm = HashAlgorithm.Create("MD5") ' Noncompliant
            Dim MD5HashAlgorithmWithNamespace = HashAlgorithm.Create("System.Security.Cryptography.MD5") ' Noncompliant
            Dim MD5Create = MD5.Create() ' Noncompliant
            Dim MD5CreateWithParam = MD5.Create("MD5") ' Noncompliant

            Dim SHA1Managed = new SHA1Managed() ' Noncompliant
            Dim SHA1Create = SHA1.Create() ' Noncompliant
            Dim SHA1CreateWithParam = SHA1.Create("SHA1") ' Noncompliant
            Dim SHA1HashAlgorithm = HashAlgorithm.Create("SHA1") ' Noncompliant
            Dim SHA1HashAlgorithmWithNamespace = HashAlgorithm.Create("System.Security.Cryptography.SHA1") ' Noncompliant
            Dim SHA1CryptoServiceProvider = new SHA1CryptoServiceProvider() ' Noncompliant

            Dim SHA256Managed = new SHA256Managed()
            Dim SHA256ManagedHashAlgorithm = HashAlgorithm.Create("SHA256Managed")
            Dim SHA256ManagedCryptoConfig = CryptoConfig.CreateFromName("SHA256Managed")

            Dim hashAlgo = HashAlgorithm.Create()

            Dim algoName = "MD5"
            Dim MD5Var = CryptoConfig.CreateFromName(algoName) ' Noncompliant

            algoName = "SHA256Managed"
            Dim SHA256ManagedVar = CryptoConfig.CreateFromName(algoName)
        End Sub

    End Class

End Namespace

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CreatingHashAlgorithms.cs
using System.Security.Cryptography;

namespace Tests.Diagnostics
{
    public class InsecureHashAlgorithm
    {
        public void Hash(byte[] temp)
        {
            using var DSACng = new DSACng(10); // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}
//                             ^^^^^^^^^^^^^^
            using var DSACryptoServiceProvider = new DSACryptoServiceProvider(); // Noncompliant
            using var DSACreate = DSA.Create(); // Noncompliant
            using var DSACreateWithParam = DSA.Create("DSA"); // Noncompliant
            using var DSACreateFromName = (AsymmetricAlgorithm)CryptoConfig.CreateFromName("DSA"); // Noncompliant
            using var DSAAsymmetricAlgorithm = AsymmetricAlgorithm.Create("DSA"); // Noncompliant

            using var HMACCreate = HMAC.Create(); // Noncompliant
            using var HMACCreateWithParam = HMAC.Create("HMACMD5"); // Noncompliant

            using var HMACMD5 = new HMACMD5(); // Noncompliant
            using var HMACMD5Create = HMACMD5.Create(); // Noncompliant
            using var HMACMD5CreateWithParam = HMACMD5.Create("HMACMD5"); // Noncompliant
            using var HMACMD5KeyedHashAlgorithm = KeyedHashAlgorithm.Create("HMACMD5"); // Noncompliant
            using var HMACMD5CryptoConfig = (HashAlgorithm)CryptoConfig.CreateFromName("HMACMD5"); // Noncompliant

            using var HMACRIPEMD160 = new HMACRIPEMD160(); // Noncompliant
            using var HMACRIPEMD160Create = HMACMD5.Create(); // Noncompliant
            using var HMACRIPEMD160CreateWithParam = HMACMD5.Create("HMACRIPEMD160"); // Noncompliant
            using var HMACRIPEMD160KeyedHashAlgorithm = KeyedHashAlgorithm.Create("HMACRIPEMD160"); // Noncompliant
            using var HMACRIPEMD160CryptoConfig = (HashAlgorithm)CryptoConfig.CreateFromName("HMACRIPEMD160"); // Noncompliant

            using var HMACSHA1 = new HMACSHA1(); // Noncompliant

            using var HMACSHA256Create = HMACSHA256.Create("HMACSHA256");
            using var HMACSHA256KeyedHashAlgorithm = KeyedHashAlgorithm.Create("HMACSHA256");
            using var HMACSHA256CryptoConfig = (HashAlgorithm)CryptoConfig.CreateFromName("HMACSHA256");

            using var MD5Cng = new MD5Cng(); // Noncompliant
            using var MD5CryptoServiceProvider = new MD5CryptoServiceProvider(); // Noncompliant
            using var MD5CryptoConfig = (HashAlgorithm)CryptoConfig.CreateFromName("MD5"); // Noncompliant
            using var MD5HashAlgorithm = HashAlgorithm.Create("MD5"); // Noncompliant
            using var MD5Create = MD5.Create(); // Noncompliant
            using var MD5CreateWithParam = MD5.Create("MD5"); // Noncompliant

            using var RIPEMD160Managed  = new RIPEMD160Managed(); // Noncompliant
            using var RIPEMD160ManagedCreate = RIPEMD160Managed.Create("RIPEMD160Managed"); // Noncompliant
            using var RIPEMD160ManagedHashAlgorithm = HashAlgorithm.Create("RIPEMD160Managed"); // Noncompliant
            using var RIPEMD160ManagedCryptoConfig = (HashAlgorithm)CryptoConfig.CreateFromName("RIPEMD160Managed"); // Noncompliant

            using var RIPEMD160Create = RIPEMD160.Create(); // Noncompliant
            using var RIPEMD160CreateWithParam = RIPEMD160.Create("RIPEMD160"); // Noncompliant
            using var RIPEMD160HashAlgorithm = HashAlgorithm.Create("RIPEMD160"); // Noncompliant
            using var RIPEMD160CryptoConfig = (HashAlgorithm)CryptoConfig.CreateFromName("RIPEMD160"); // Noncompliant

            using var SHA1Create = SHA1.Create(); // Noncompliant
            using var SHA1CreateWithParam = SHA1.Create("SHA1"); // Noncompliant
            using var SHA1Managed = new SHA1Managed(); // Noncompliant
            using var SHA1HashAlgorithm = HashAlgorithm.Create("SHA1"); // Noncompliant
            using var SHA1CryptoServiceProvider = new SHA1CryptoServiceProvider(); // Noncompliant

            using var sha256Managed = new SHA256Managed();
            using var sha256HashAlgorithm = HashAlgorithm.Create("SHA256Managed");
            var sha256CryptoConfig = CryptoConfig.CreateFromName("SHA256Managed");

            HashAlgorithm hashAlgo = HashAlgorithm.Create();

            var algoName = "MD5";
            var md5Var = (HashAlgorithm)CryptoConfig.CreateFromName(algoName); // Noncompliant
            algoName = "SHA256Managed";
            var SHA256ManagedVar = (HashAlgorithm)CryptoConfig.CreateFromName(algoName);
        }
    }
}
